### PR TITLE
fix hostname detection logic

### PR DESCRIPTION
**Pull request by xvello**
### What does this PR do?

Rework the hostname detection logic.

### Motivation

`master` currently fails on **dockerized non-ECS EC2** hosts because the error handling is inconsistent between the different providers.
This PR reworks the detection logic by iterating over an array to make sure all providers have the same error handling.

Also, validate all hostnames with `ValidHostname` to handle EC2 returning `"", nil` sometimes. We can now put it on the first bucket, with GCE, for consistent cross-cloud behaviour.

# Testing

## GCE host ✅
```
xaviervello@gke-xvello-test-18-default-pool-3f300d61-548w ~ $ docker pull datadog/agent-dev:hostname && docker run --rm -e DD_LOG_LEVEL=debug -e DD_API_KEY=bla datadog/agent-
2017-11-24 13:31:43 UTC | INFO | (hostname.go:107 in lookupHostname) | unable to get the hostname from the config file: host name is empty
2017-11-24 13:31:43 UTC | INFO | (hostname.go:108 in lookupHostname) | trying to determine a reliable host name automatically...
2017-11-24 13:31:43 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from ec2...
2017-11-24 13:31:43 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from ec2: unable to fetch EC2 API, status code 404 trying to fetch http://169.254.169.254/latest/meta-data/instance-id
2017-11-24 13:31:43 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from gce...
2017-11-24 13:31:43 UTC | INFO | (hostname.go:123 in lookupHostname) | got hostname "gke-xvello-test-18-default-pool-3f300d61-548w.c.datadog-demo-1336.internal" from gce
```
## EC2 host ✅
```
ubuntu@ip-172-31-18-78:~$ docker pull datadog/agent-dev:hostname && docker run --rm -e DD_LOG_LEVEL=debug -e DD_API_KEY=bla datadog/agent-dev:hostname|grep hostname.go
2017-11-24 13:30:19 UTC | INFO | (hostname.go:107 in lookupHostname) | unable to get the hostname from the config file: host name is empty
2017-11-24 13:30:19 UTC | INFO | (hostname.go:108 in lookupHostname) | trying to determine a reliable host name automatically...
2017-11-24 13:30:19 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from ec2...
2017-11-24 13:30:19 UTC | INFO | (hostname.go:123 in lookupHostname) | got hostname "i-073450472af8fe2b8" from ec2
```
## Local vagrant in docker ✅
```
ubuntu@ci-xaviervello:~/go/src/github.com/DataDog/datadog-agent$ docker run --rm -v /var/run/docker.sock:/var/run/docker.sock -e DD_LOG_LEVEL=debug -e DD_API_KEY=bla datadog/agent-dev:hostname|grep hostname.go
2017-11-24 13:33:53 UTC | INFO | (hostname.go:107 in lookupHostname) | unable to get the hostname from the config file: host name is empty
2017-11-24 13:33:53 UTC | INFO | (hostname.go:108 in lookupHostname) | trying to determine a reliable host name automatically...
2017-11-24 13:33:53 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from ec2...
2017-11-24 13:33:53 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from ec2: unable to fetch EC2 API, Get http://169.254.169.254/latest/meta-data/instance-id: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2017-11-24 13:33:53 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from gce...
2017-11-24 13:33:53 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from gce: unable to retrieve hostname from GCE: Get http://169.254.169.254/computeMetadata/v1/instance/hostname: dial tcp 169.254.169.254:80: getsockopt: no route to host
2017-11-24 13:33:53 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from docker...
2017-11-24 13:33:53 UTC | INFO | (hostname.go:123 in lookupHostname) | got hostname "ci-xaviervello" from docker
```

## Local vagrant in docker no access to socket 🛑 so ✅
```
ubuntu@ci-xaviervello:~/go/src/github.com/DataDog/datadog-agent$ docker run --rm -e DD_LOG_LEVEL=debug -e DD_API_KEY=bla datadog/agent-dev:hostname|grep hostname.go
2017-11-24 13:32:58 UTC | INFO | (hostname.go:107 in lookupHostname) | unable to get the hostname from the config file: host name is empty
2017-11-24 13:32:58 UTC | INFO | (hostname.go:108 in lookupHostname) | trying to determine a reliable host name automatically...
2017-11-24 13:32:58 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from ec2...
2017-11-24 13:32:58 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from ec2: unable to fetch EC2 API, Get http://169.254.169.254/latest/meta-data/instance-id: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2017-11-24 13:32:58 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from gce...
2017-11-24 13:32:58 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from gce: unable to retrieve hostname from GCE: Get http://169.254.169.254/computeMetadata/v1/instance/hostname: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2017-11-24 13:32:58 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from docker...
2017-11-24 13:32:58 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from docker: temporary failure in dockerutil, will retry later: docker not available
Error: Error while getting hostname, exiting: Unable to reliably determine the host name. You can define one in the agent config file or in your hosts file
```
## Local vagrant, on the host, docker running ✅
```
ubuntu@ci-xaviervello:~/go/src/github.com/DataDog/datadog-agent$ ./bin/agent/agent -c ./bin/agent/dist/ start |grep hostname.go
2017-11-24 13:32:17 UTC | INFO | (hostname.go:107 in lookupHostname) | unable to get the hostname from the config file: host name is empty
2017-11-24 13:32:17 UTC | INFO | (hostname.go:108 in lookupHostname) | trying to determine a reliable host name automatically...
2017-11-24 13:32:17 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from ec2...
2017-11-24 13:32:17 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from ec2: unable to fetch EC2 API, Get http://169.254.169.254/latest/meta-data/instance-id: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2017-11-24 13:32:17 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from gce...
2017-11-24 13:32:17 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from gce: unable to retrieve hostname from GCE: Get http://169.254.169.254/computeMetadata/v1/instance/hostname: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2017-11-24 13:32:17 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from docker...
2017-11-24 13:32:17 UTC | INFO | (hostname.go:123 in lookupHostname) | got hostname "ci-xaviervello" from docker
```

## Local vagrant, on the host, docker stopped ✅
```
ubuntu@ci-xaviervello:~/go/src/github.com/DataDog/datadog-agent$ ./bin/agent/agent -c ./bin/agent/dist/ start |grep hostname.go
2017-11-24 13:07:41 UTC | INFO | (hostname.go:107 in lookupHostname) | unable to get the hostname from the config file: host name is empty
2017-11-24 13:07:41 UTC | INFO | (hostname.go:108 in lookupHostname) | trying to determine a reliable host name automatically...
2017-11-24 13:07:41 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from ec2...
2017-11-24 13:07:41 UTC | DEBUG | (hostname.go:120 in lookupHostname) | invalid hostname from ec2: host name is empty
2017-11-24 13:07:41 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from gce...
2017-11-24 13:07:42 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from gce: unable to retrieve hostname from GCE: Get http://169.254.169.254/computeMetadata/v1/instance/hostname: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2017-11-24 13:07:42 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from docker...
2017-11-24 13:07:42 UTC | DEBUG | (hostname.go:115 in lookupHostname) | cannot get hostname from docker: unable to get Docker info: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
2017-11-24 13:07:42 UTC | DEBUG | (hostname.go:112 in lookupHostname) | trying to get hostname from OS...
2017-11-24 13:07:42 UTC | INFO | (hostname.go:123 in lookupHostname) | got hostname "ci-xaviervello" from OS
```

### CODE REVIEW

File: pkg/util/ec2/ec2.go [hunk 1]
```diff
@@ -83,9 +80,5 @@ func IsDefaultHostname(hostname string) bool {
 
 // HostnameProvider gets the hostname
 func HostnameProvider(hostName string) (string, error) {
-	if ecs.IsInstance() || IsDefaultHostname(hostName) {
```
Comment (JulienBalestra) [inline pkg/util/ec2/ec2.go]: Maybe we can change the signature of HostnameProvider(s) from
`func HostnameProvider(hostName string) (string, error)` to `func HostnameProvider() (string, error)` ?

File: pkg/util/hostname.go [hunk 2]
```diff
@@ -80,76 +80,51 @@ func Fqdn(hostname string) string {
...
-			return name, err
-		}
-	}
-
-	isContainerized, name := getContainerHostname()
-	if isContainerized && name != "" {
-		hostName = name
-	}
+	log.Infof("unable to get the hostname from the config file: %s", err)
+	log.Info("trying to determine a reliable host name automatically...")
```
Comment (JulienBalestra) [inline pkg/util/hostname.go]: We can make a simple call of the log for both lines.

File: pkg/util/hostname.go [hunk 3]
```diff
@@ -80,76 +80,51 @@ func Fqdn(hostname string) string {
...
+			if err != nil {
+				log.Debugf("cannot get hostname from %s: %s", provider.Name, err)
+				continue
+			}
+			err = ValidHostname(name)
+			if err != nil {
+				log.Debugf("invalid hostname from %s: %s", provider.Name, err)
+				continue
+			}
+			log.Infof("got hostname \"%s\" from %s", name, provider.Name)
```
Comment (JulienBalestra) [inline pkg/util/hostname.go]: ditto for `%q`

File: pkg/util/hostname.go [hunk 4]
```diff
@@ -80,76 +80,51 @@ func Fqdn(hostname string) string {
...
-	}
-
-	cache.Cache.Set(cacheHostnameKey, hostName, cache.NoExpiration)
-	return hostName, err
-}
-
-// IsKubernetes returns whether the Agent is running on a kubernetes cluster
-func isKubernetes() bool {
-	return os.Getenv("KUBERNETES_PORT") != ""
+	return "", fmt.Errorf("Unable to reliably determine the host name. You can define one in the agent config file or in your hosts file")
```
Comment (JulienBalestra) [inline pkg/util/hostname.go]: s/Unable/unable

**Review by JulienBalestra — CHANGES_REQUESTED** [review]
lgtm, a few tricks for the quote formatting and see if we remove the string parameter in the `HostnameProvider`

File: pkg/util/hostname.go [hunk 5]
```diff
@@ -80,76 +80,51 @@ func Fqdn(hostname string) string {
...
+	log.Info("trying to determine a reliable host name automatically...")
 
-	if hostName == "" {
-		// os
-		log.Debug("GetHostname trying os...")
-		name, err = os.Hostname()
-		if err == nil {
-			hostName = name
+	for _, providers := range hostname.ProviderCatalog {
+		for _, provider := range providers {
```
Comment (tmichelet) [inline pkg/util/hostname.go]: I would hardcode the list of providers, so we keep the order constant

File: pkg/util/hostname/providers.go [hunk 6]
```diff
@@ -5,13 +5,33 @@
...
 
-// Provider is a generic function to grab the hostname and return it
-type Provider func(string) (string, error)
+// ProviderMethod is a generic function to grab the hostname and return it
+type ProviderMethod func(string) (string, error)
 
-// ProviderCatalog holds all the various kinds of hostname providers
-var ProviderCatalog = make(map[string]Provider)
+// ProviderPriority allows to specify a lookup priority for a provider
+type ProviderPriority uint
```
Comment (hkaj) [inline pkg/util/hostname/providers.go]: This is way overkill. As @tmichelet mentioned above we should probably just hardcode providers.

**Review by hkaj — CHANGES_REQUESTED** [review]

**Pull request closed without merging**